src/toil/job.py

```diff
@@ -59,246 +59,897 @@
+            if self.command is not None:
+                assert self.command == self.checkpoint
+                logger.debug("Checkpoint job already has command set to run")
+            else:
+                self.command = self.checkpoint
+
+            jobStore.update(self) # Update immediately to ensure that checkpoint
+            # is made before deleting any remaining successors
+
+            if self.childIDs or self.followOnIDs or self.serviceTree:
+                # If the subtree of successors is not complete restart everything
+                logger.debug("Checkpoint job has unfinished successor jobs, deleting children: %s, followOns: %s, services: %s " %
+                             (self.childIDs, self.followOnIDs, self.serviceTree.keys()))
+                # Delete everything on the stack, as these represent successors to clean
+                # up as we restart the queue
+                def recursiveDelete(jobDesc):
+                    # Recursive walk the stack to delete all remaining jobs
+                    for otherJobID in jobDesc.successorsAndServiceHosts():
+                        if jobStore.exists(otherJobID):
+                            recursiveDelete(jobStore.load(otherJobID))
+                        else:
+                            logger.debug("Job %s has already been deleted", otherJobID)
+                    if jobDesc.jobStoreID != self.jobStoreID:
+                        # Delete everything under us except us.
+                        logger.debug("Checkpoint is deleting old successor job: %s", jobDesc.jobStoreID)
+                        jobStore.delete(jobDesc.jobStoreID)
+                        successorsDeleted.append(jobDesc.jobStoreID)
+                recursiveDelete(self)
+                
+                # Cut links to the jobs we deleted.
+                self.clearSuccessorsAndServiceHosts()
 
-    @classmethod
-    def fromJob(cls, job, command, predecessorNumber):
-        """
-        Build a job node from a job object
-        :param toil.job.Job job: the job object to be transformed into a job node
-        :param str command: the JobNode's command
-        :param int predecessorNumber: the number of predecessors that must finish
-            successfully before the job can be scheduled
-        :return: a JobNode object representing the job object parameter
-        :rtype: toil.job.JobNode
-        """
-        return cls(jobStoreID=None,
-                   requirements=job._requirements,
-                   command=command,
-                   jobName=job.jobName,
-                   unitName=job.unitName,
-                   displayName=job.displayName,
-                   predecessorNumber=predecessorNumber)
-
-class Job(BaseJob):
+                # Update again to commit the removal of successors.
+                jobStore.update(self)
+        return successorsDeleted
+
+class Job:
```

Comment:
The way it works is that both `Job` and `JobDescription` have `cores` and friends as properties. `JobDescription` stores the actual values, and `Job` fulfills all reads and writes by sending them along to the associated `JobDescription`.

The values come in initially through the `__init__` of Job, and they're passed along to the JobDescription when it is created.

Note that descriptionClass is a class, *not* a JobDescription instance. Job is responsible for creating its own JobDescription. It's just that some Job descendants need to use JobDescription descendants as their descriptions, and the best way I thought of to do that was to pass the type along to Job's `__init__`. Otherwise all the name determining and requirements munging would have to move to methods that all the Job descendants would have to call in their `__init__` methods before constructing their descriptions.